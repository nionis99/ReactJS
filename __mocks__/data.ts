export const genres = ['Crime', 'Documentary', 'Horror', 'Comedy'];
export const MovieMockData = {
  id: 0,
  title: 'The Avengers 2',
  vote_average: 7.5,
  release_date: '2012-04-25',
  poster_path: 'https://image.tmdb.org/t/p/w500/cezWGskPY5x7GaglTTRN4Fugfb8.jpg',
  overview: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
  genres: ['Science Fiction', 'Action', 'Adventure'],
  runtime: 143,
};
