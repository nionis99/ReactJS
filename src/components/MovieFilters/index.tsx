import React, { HTMLAttributes } from 'react';
import cx from 'classnames';

import ArrowDown from 'assets/icons/arrowDown.svg';
import useQuery from 'hooks/useQuery';

const defaultTabClassName =
  'text-white p-1 md:p-5 cursor-pointer hover:text-primary focus:outline-none uppercase font-medium border-b-2';

const sortOptions = [
  { name: 'Genres', value: 'genres' },
  { name: 'Rating', value: 'vote_average' },
  { name: 'Release date', value: 'release_date' },
];

const MovieFilters = ({ className }: HTMLAttributes<HTMLDivElement>) => {
  const { currentQuery, addQuery } = useQuery();
  const genreFilter = currentQuery.get('genre');
  const sortByValue = currentQuery.get('sortBy');

  const onSortChange = (event: React.ChangeEvent<HTMLSelectElement>) => addQuery('sortBy', event.target.value);

  const onFilterClick = (event: React.MouseEvent<HTMLSpanElement>) => {
    const selectedFilter = event.currentTarget.innerText[0] + event.currentTarget.innerText.substring(1).toLowerCase();
    addQuery('genre', selectedFilter);
  };

  return (
    <nav className={cx('flex flex-row overflow-auto', className)}>
      <span
        className={cx(defaultTabClassName, genreFilter === 'All' ? 'border-primary text-primary' : 'text-white')}
        onClick={onFilterClick}
      >
        All
      </span>
      <span
        className={cx(
          defaultTabClassName,
          genreFilter === 'Documentary' ? 'border-primary text-primary' : 'text-white',
        )}
        onClick={onFilterClick}
      >
        Documentary
      </span>
      <span
        className={cx(defaultTabClassName, genreFilter === 'Comedy' ? 'border-primary text-primary' : 'text-white')}
        onClick={onFilterClick}
      >
        Comedy
      </span>
      <span
        className={cx(defaultTabClassName, genreFilter === 'Horror' ? 'border-primary text-primary' : 'text-white')}
        onClick={onFilterClick}
      >
        Horror
      </span>
      <span
        className={cx(defaultTabClassName, genreFilter === 'Crime' ? 'border-primary text-primary' : 'text-white')}
        onClick={onFilterClick}
      >
        Crime
      </span>
      <span className="cursor-default mx-auto w-full border-b-2" />
      <span className="text-white p-1 md:p-5 uppercase font-normal border-b-2 text-opacity-50 whitespace-nowrap">
        Sort by
      </span>
      <div className="relative">
        <select
          className="appearance-none w-auto bg-transparent text-white p-1 md:p-5 pr-4 border-b-2 font-medium
          focus:outline-none truncate cursor-pointer"
          onChange={onSortChange}
          defaultValue={sortByValue || 'genres'}
        >
          {sortOptions.map((sortOption, index) => (
            <option key={index} value={sortOption.value}>
              {sortOption.name}
            </option>
          ))}
        </select>
        <div className="pointer-events-none absolute inset-y-0 -right-2 flex items-center px-2 text-gray-700">
          <ArrowDown className="fill-current h-3 w-3" />
        </div>
      </div>
    </nav>
  );
};

export default MovieFilters;
