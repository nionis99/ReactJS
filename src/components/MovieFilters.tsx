import React, { HTMLAttributes } from 'react';
import cx from 'classnames';
import ArrowDown from 'assets/icons/arrowDown.svg';
import { sortOptions } from '../../__mocks__/data';
import useMoviePageContext from 'contexts/MoviePageProvider';

const defaultTabClassName =
  'text-white p-1 md:p-5 cursor-pointer hover:text-primary focus:outline-none uppercase font-medium border-b-2';

const MovieFilters = ({ className }: HTMLAttributes<HTMLDivElement>) => {
  const { setFilter, setSortBy, sortBy, filter } = useMoviePageContext();
  console.log(filter, sortBy);
  const onSortChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setSortBy(event.target.value);
  };

  const onFilterClick = (event: React.MouseEvent<HTMLSpanElement>) => {
    const selectedFilter = event.currentTarget.innerText[0] + event.currentTarget.innerText.substring(1).toLowerCase();
    selectedFilter === 'All' ? setFilter(null) : setFilter(selectedFilter);
  };

  return (
    <nav className={cx('flex flex-row overflow-auto', className)}>
      <span
        className={cx(defaultTabClassName, !filter ? 'border-primary text-primary' : 'text-white')}
        onClick={onFilterClick}
      >
        All
      </span>
      <span
        className={cx(defaultTabClassName, filter === 'Documentary' ? 'border-primary text-primary' : 'text-white')}
        onClick={onFilterClick}
      >
        Documentary
      </span>
      <span
        className={cx(defaultTabClassName, filter === 'Comedy' ? 'border-primary text-primary' : 'text-white')}
        onClick={onFilterClick}
      >
        Comedy
      </span>
      <span
        className={cx(defaultTabClassName, filter === 'Horror' ? 'border-primary text-primary' : 'text-white')}
        onClick={onFilterClick}
      >
        Horror
      </span>
      <span
        className={cx(defaultTabClassName, filter === 'Crime' ? 'border-primary text-primary' : 'text-white')}
        onClick={onFilterClick}
      >
        Crime
      </span>
      <span className="cursor-default mx-auto w-full border-b-2" />
      <span className="text-white p-1 md:p-5 uppercase font-normal border-b-2 text-opacity-50 whitespace-nowrap">
        Sort by
      </span>
      <div className="relative">
        <select
          className="appearance-none w-auto bg-transparent text-white p-1 md:p-5 pr-4 border-b-2 font-medium
          focus:outline-none truncate cursor-pointer"
          onChange={onSortChange}
          defaultValue={sortBy || 'genres'}
        >
          {sortOptions.map((sortOption, index) => (
            <option key={index} value={sortOption.value}>
              {sortOption.name}
            </option>
          ))}
        </select>
        <div className="pointer-events-none absolute inset-y-0 -right-2 flex items-center px-2 text-gray-700">
          <ArrowDown className="fill-current h-3 w-3" />
        </div>
      </div>
    </nav>
  );
};

export default MovieFilters;
