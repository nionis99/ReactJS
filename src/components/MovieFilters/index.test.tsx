import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { fireEvent, render } from '@testing-library/react';
import { genres } from '../../../__mocks__/data';
import MovieFilters from '.';

describe('Movie filters', () => {
  const wrapper = (
    <BrowserRouter>
      <MovieFilters />
    </BrowserRouter>
  );

  it('should render movie filters component', () => {
    const elem = renderer.create(wrapper).toJSON();
    expect(elem).toMatchSnapshot();
  });

  it('after selected genre it should be highlighted', () => {
    const { getByText } = render(wrapper);
    console.log(getByText(genres[0]).className);
    fireEvent.click(getByText(genres[0]));
    console.log(getByText(genres[0]).className);
    // expect(getByText(genres[0]).className).toContain('text-primary border-primary');
  });
});
