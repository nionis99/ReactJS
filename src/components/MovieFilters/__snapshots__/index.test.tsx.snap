// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie filters should render movie filters component 1`] = `
<nav
  className="flex flex-row overflow-auto"
>
  <span
    className="text-white p-1 md:p-5 cursor-pointer hover:text-primary focus:outline-none uppercase font-medium border-b-2 text-white"
    onClick={[Function]}
  >
    All
  </span>
  <span
    className="text-white p-1 md:p-5 cursor-pointer hover:text-primary focus:outline-none uppercase font-medium border-b-2 text-white"
    onClick={[Function]}
  >
    Documentary
  </span>
  <span
    className="text-white p-1 md:p-5 cursor-pointer hover:text-primary focus:outline-none uppercase font-medium border-b-2 text-white"
    onClick={[Function]}
  >
    Comedy
  </span>
  <span
    className="text-white p-1 md:p-5 cursor-pointer hover:text-primary focus:outline-none uppercase font-medium border-b-2 text-white"
    onClick={[Function]}
  >
    Horror
  </span>
  <span
    className="text-white p-1 md:p-5 cursor-pointer hover:text-primary focus:outline-none uppercase font-medium border-b-2 text-white"
    onClick={[Function]}
  >
    Crime
  </span>
  <span
    className="cursor-default mx-auto w-full border-b-2"
  />
  <span
    className="text-white p-1 md:p-5 uppercase font-normal border-b-2 text-opacity-50 whitespace-nowrap"
  >
    Sort by
  </span>
  <div
    className="relative"
  >
    <select
      className="appearance-none w-auto bg-transparent text-white p-1 md:p-5 pr-4 border-b-2 font-medium
          focus:outline-none truncate cursor-pointer"
      defaultValue="genres"
      onChange={[Function]}
    >
      <option
        value="genres"
      >
        Genres
      </option>
      <option
        value="vote_average"
      >
        Rating
      </option>
      <option
        value="release_date"
      >
        Release date
      </option>
    </select>
    <div
      className="pointer-events-none absolute inset-y-0 -right-2 flex items-center px-2 text-gray-700"
    >
      <test-file-stub
        className="fill-current h-3 w-3"
      />
    </div>
  </div>
</nav>
`;
